In [1]:
from __future__ import division, print_function

# 9. Python functions 
________

Sometimes, a portion of code is reused over and over again in the entire script. To prevent repetitive coding, we are able to define our own custom defined functions using the `def` keyword. When invoked, functions will instruct the computer to perform a set list of instructions, possibly returning an output at the end. Functions can also be pre-defined and saved in another file (with extension `.py`) so that it can be used for another project. 

You will have encountered many pre-defined functions in Python up to now. Functions like `print`, `len`, etc... But there are many other pre-built functions in Python which are extremely useful and makes code more transparent and readable.

We will also encounter what is known as anonymous functions which are defined using the `lambda` keyword. These are short functions and can only be defined in one line of code. While technically `def` does what `lambda` does, nevertheless `lambda` allows for less pendantic and more natural style of coding. You will encounter anonymous functions alot when working with `pandas` especially when cleaning up data frames. 

Finally we introduce what is perhaps one of the more important concepts in Python programming: List comprehensions. These are syntatical shortcuts for the `for` loop which translate not only to better code style, but faster scripts. 

## 9.1 Learning objectives 

The objectives of this unit are:

1. To use the `def` keyword to defined functions. 

1. Recall the terms  arguments, keyword arguments, the signature of a function. 

1. To use `lambda` to define anonymous functions. 

1. To use build in  functions: `zip` and `enumerate`.  

1. To learn how to refactor `for` loops into list comprehension statements. 

## 9.2 Let's build our own functions

All functions are defined using the `def` keyword. In general, the format of function definition looks like this: 

    def my_function_name(arg_1, arg_2, ..., arg_n) : 
    
        code
        
        return <something, or None> 
        
First, we tell Python that we are going to define a function by typing out `def`. Then we proceed by naming our function. The normal rules for naming variables apply to naming functions as well - one cannot start a name with a number or use special characters or use words which have been reserved for Python. 

Then after the name, we describe the *signature* of the function by writing down every argument to the function seperated by commas and enclosed in round braces `( )`. And *argument* to a function is an input to the code which will be executed when the function is *called*. It is not mandatory that a function recieve inputs. Sometimes, a function just needs to run a set of instructions, without any input. We end the `def` statement with a `:`. The newline under this marks the beginning of the function code. 

Every line of code meant for the function *must* be indented. There are no enclosing `{ }` which marks the "body" of the function. In Python, the "body" of the function is denoted with indentation only. Thus, every line of code meant for the function must be on the same indentation level. Finally, at the end of the function we `return` an output or `None`. While this last syntax is not mandatory, it is not good practice to leave off a function definition *without* a `return` statement.  

In [2]:
# Our first function

def my_first_function():
    pass

For our first function, we see above that `my_first_function` does not take in any input and does nothing. The `pass` keyword is a kind of temporary placeholder and basically does nothing. We use `pass` because one cannot leave a function "body" without any code at all. 

Now let's code something into `my_first_function` so that it does something useful. 

In [3]:
def my_first_function():
    print("Hello world!")

`my_first_function` will `print` the string `"Hello world!"` whenever it is *called*. Calling a function basically means instructing Python to run the code contained in the function. Notice that after defining a function and running the cell, there is no output. But that doesn't mean nothing has happened. In fact, Python has populated the global namespace with a new name, `my_first_function` and is ready to do what ever has been coded into this function when it is called. 

In [4]:
my_first_function

<function __main__.my_first_function>

It is good to understand what happens when we type `my_first_function` and execute a cell. Notice that the output says `<function ...` This means that the variable `my_first_function` represents an object of type `function`. The rest of the output indicates that this function is represented by a name `my_first_function` in the module `__main__`. We will not describe what modules are in this course, but suffices for our purposes to think of `__main__` as file containing all the functions that we will define in this Jupyter Notebook session. 

To actually execute the instructions in `my_first_function`, we must type `my_first_function()`. 

In [5]:
my_first_function()

Hello world!


## 9.3 Functions with arguments

Functions won't be useful if we are unable to pass input into it. Most of the time, the set of instructions will act on the input we have supplied to the function and produces some output which is then passed to a variable to be stored. Let's modify `my_first_function` to print out a name supplied as input to it. 

In [6]:
def my_first_function(name):
    print("Hello %s" % (name))
    return None

my_first_function("Tang U-Liang")

Hello Tang U-Liang


In [15]:
# Passing two arguments

def special_product(x,y):
    prod = x-y+x*y
    return prod 

When defining functions with arguments, the same variable name used in the signature must be used in the body of the function. Now there is nothing inherently special about using `name` to represent the argument for names to `my_first_function`. After all, the computer doesn't "understand" that we intend to print out a name when calling `my_first_function`. However, we should use recognizable variable names to improve readibility of our code and to make our intentions transparent. 

In the function `special_product`, I passed two arguments named `x` and `y`. Inside the function, it performs the operation and assigns the result to a variable named `prod`. Then the function uses the keyword `return` to send the answer out from the function environment to the global environment. 

In [16]:
answer = special_product(1,3)
print(answer)

1


What  happened is that the function `special_product` performs the said operation on inputs `1` and `3`. It then outputs the answer, in this case `7`. We assign the output `7` to a variable named `answer` and print it. 

Note that we do not need to explicit declare a variable to "capture" the answer. The following works too. 

In [17]:
special_product(1,3)

1

Passing arguments in correct sequence matters. Python will pass values to arguments according to the sequence as it was declared in the signature. 

In [20]:
# x = 1, and y =3
print(special_product(1,3))

# x =3 and y = 1
print(special_product(3,1))

1
5


What will happen if we try to display the variable `prod` directly? 

In [18]:
print(prod)

NameError: name 'prod' is not defined

### 9.3.1 Function scope

But isn't the variable `prod` defined already when we defined the function `special_product`? This happens because the variable `prod` is only available in the *scope* of the function. The global environment is another scope. In general, variables from one scope are not accessible in another scope with exeptions given by scoping rules (which are programming language dependant). For this course, it suffices to know that variables defined in the function scope will *NOT* be accessible from the global scope. 

(This can be overriden using the `global` keyword. But this is not encouraged.) 

## 9.4 Function defaults

When we define functions, all variables we define in the signature must be assigned values. We cannot leave any out. 

In [21]:
special_product(1,)

TypeError: special_product() takes exactly 2 arguments (1 given)

Therefore, it becomes quite a hassle if we have to call the function in various places in our code with the same input in one of the arguments. To do that we can assign default values to particular arguments in the following manner.

    def my_function(arg_1, arg_2 = default_value, ...):
    
        code
        
Note that arguments assigned default values must come after arguments without default values. Also, don't worry that you cannot input values other than defaults. You are still able to override default values when you need to. 

In [24]:
def special_product(x, y=1): # default value of y is 1
    return x-y+x*y

# We don't have to pass any value to arguments with default values
print(special_product(2))

# Default values can be overriden
print(special_product(2,9))

3
11


## 9.5 Passing arguments to functions by keyword 

The arguments to a function have names, just as variables have names. The names of arguments to a function are called keywords. We can pass arguments to functions by assigning values explicitly to keywords like so:

    my_function(keyword_1 = value_1, keyword_2 = value_2,...)
    
This gives enormous flexibility in using Python interactively. Most functions given in the `matplotlib` and `seaborn` libraries have many arguments almost all of have them have default values. However, we often use a few of these keywords and it is quite a pain to remember the exact sequence of arguments in the function signature. Passing argments to keywords allows us to pass arguments in any order convenient to us. 

In [25]:
special_product(1,3) == special_product(3,1)

False

In [27]:
special_product(x=1, y=3) == special_product( y=3, x=1)

True

### 9.5.1 An application: A function to implement the Euclidean algorithm

To end this section, below is code to find the gcd of two positive integers. The function will also print out each stage of the algorithm as an output of 3 numbers: (m, n, remainder). 

In [1]:
def gcd(m=1,n=1):
    """
    This function implementats the Euclidean algorithm to find the gcd of two positive integers. 
    
    Returns:
        int, greatest common divisor of m and n. 
    """
    
    if m < n:
        r = [n, m, n%m]
    else:
        r = [m, n, m%n]
        
    for _ in range(0, r[1]):
        
        print("{} {} {}".format(r[0], r[1], r[2]))
        
        if r[2] == 0:
            break
        else:
            r = [r[1], r[2], r[1]%r[2]]
        
    print("Greatest common divisor of {} and {} is {}".format(m, n, r[1]))
    return r[1]

In [2]:
gcd(151000,14356)

151000 14356 7440
14356 7440 6916
7440 6916 524
6916 524 104
524 104 4
104 4 0
Greatest common divisor of 151000 and 14356 is 4


4

# 10. Lambda expressions
_______

Lambda expressions are used to define short functions that may be written in one line of code. This is more than just a convenience. Most arguments to `pandas` and `seaborn` functions are intended to take in *callables* (functions) and lambda expressions provide a good syntactical way of passing functions as arguments to other functions. 

Recall our definition of `my_first_function`: 

In [3]:
def my_first_function(name):
    print("Hello %s" % (name))

Notice that this function essentially consists of one line, namely the `print` statement. Using lambda expressions, this can be shortened to:

In [8]:
printer = lambda name: print("Hello %s" % (name))

We use the `lambda` keyword to define lambda expressions. After `lambda` we type in the arguments to the function but without enclosing it in `( )`. All arguments must be seperated by commas. Once that is done, type a `:` and follow it with one line of code which does whatever you want it to do. In this case here, I simply want to `print` a name. Functionally, this lambda expression is equivalent to `my_first_function`. However, as you can see below, they are different objects.

In [9]:
my_first_function is printer

False

In [10]:
printer

<function __main__.<lambda>>

Notice that `printer` is of class `function` but is given a name `<lambda>`. However, we can call `printer` just as we called `my_first_function`, by passing arguments to it.  

In [11]:
printer("Joe")

Hello Joe


Lambda expressions can take on more than one argument. Here is the function `special_product` refactored as a lambda expression.

In [15]:
special_product = lambda x, y: x-y+x*y

special_product(10,9)

91

Notice that I did not need to put a `return` to indicate which output to pass to the global environment. This is because lambda expressions are meant to be written in one line, hence it is understood that that one line of code *is* the output. 

## 10.1 Use cases

Lambda expressions are also known as anonymous functions because we rarely assign lambda expressions to variables. Instead, they are passed directly to keywords or as arguments to most `pandas` functions or methods. Here is an example to how this is used in a pandas dataframe.

In what follows, we intend to calculate the ratio of `sulphates` to `alchohol` content for each sample (row) and assign it as a new column to the data frame. The data frame is displayed below and has been assigned to variable named `wine`. 

In [17]:
import pandas as pd # Importing the pandas library

wine = pd.read_csv("winequality-red.csv", sep=';')

wine.sample(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
290             8.7             0.520         0.09             2.5      0.091   
1478            7.1             0.875         0.05             5.7      0.082   
622            10.0             0.580         0.22             1.9      0.080   
287             7.5             0.400         0.12             3.0      0.092   
1292            5.9             0.395         0.13             2.4      0.056   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
290                  20.0                  49.0  0.99760  3.34       0.86   
1478                  3.0                  14.0  0.99808  3.40       0.52   
622                   9.0                  32.0  0.99740  3.13       0.55   
287                  29.0                  53.0  0.99670  3.37       0.70   
1292                 14.0                  28.0  0.99362  3.62       0.67   

      alcohol  quality  
290      10.6        7  
1478     10.2        3  
622       9.5        5  
287      10.3        6  
1292     12.4        6

Here's how this could be achieved. We first define the function that calculates the ratio and then proceed to create the new calculated column. 

In [20]:
def ratio(df): 
    """ This function calculates the ratio of sulphates to alcohol content in the wine dataframe
    
    Returns
        Series, shape (n_samples, ) Array containing the ratio of sulphate to alcohol content for each sample
    """
    ratio_col = df.sulphates/df.alcohol
    return ratio_col

(wine.assign(ratio_sul_to_alc=ratio)
     .head(5))

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  ratio_sul_to_alc  
0      9.4        5          0.059574  
1      9.8        5          0.069388  
2      9.8        5          0.066327  
3      9.8        6          0.059184  
4      9.4        5          0.059574

As you can see, a new column has been added with the calculated column named `ratio_sul_to_alc`. However, we had to define a function named `ratio` which we may or may not use again. We would like to achieve the same thing, but without populating the global namespace with unnecessary variables.  

So let's do the same thing but with lambda expressions. 

In [21]:
(wine.assign(ratio_sul_to_alc=lambda df: df.sulphates/df.alcohol)
     .head(5))

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  ratio_sul_to_alc  
0      9.4        5          0.059574  
1      9.8        5          0.069388  
2      9.8        5          0.066327  
3      9.8        6          0.059184  
4      9.4        5          0.059574

Notice that they give the same answer. We will learn how to do this in detail in the next unit. For now, the purpose of this example is to illustrate how lambda expressions are a great help in simplifying and making code more compact and readable. 

# 11. Built in functions
_____
Python has some built in functions for coding purposes. While there are quite a few of them, the following three will be used quite often in handling dataframes and visualization. These are 

  * `zip` is a utility function that produces tuples from two lists of equal length. 
  * `enumerate` also creates a tuple in the form $ (i, item_i)$ for $0\leq i < $ `len(items)`. 

We will also cover the concept of **list comprehension** in this section as it is an important programming concept and syntax in Python. 

## 11.1 `zip`

To understand what `zip` does, we need to describe a rather simple data structure called `tuple`. A tuple is like a list, with the difference being that its elements are **immutable**. Tuples are created by enclosing a list of objects seperated by commas within two round braces `()`

In [1]:
pair = (1,4)

print(pair)

(1, 4)


As with lists, tuples can also be indexed and sliced. However, once assigned, individual components of a tuple cannot be changed. For example, the following code will raise and error

In [4]:
pair[0] = 2

TypeError: 'tuple' object does not support item assignment

Think of tuples as lists which you wish to protect from changing by accidental assignment. Another way of thinking about tuples are also as constant lists, or as "coordinates" in $\mathbb{R}^n$. 

Given two lists with items $$ x_1, \ldots, x_n$$ and $$ y_1, \ldots, y_n$$ `zip` produces a new list of `tuples` in the following form $$ (x_1, y_1), \ldots, (x_n, y_n)$$

To understand how `zip` works, let's try to replicate its function using a for loop.  

In [6]:
zipped = list() # This creates and empty list 

my_colleagues = ['Andy', 'Lisa', 'Dayton']
ages = [29, 24, 50]

for i in range(0,3):
    zipped.append((my_colleagues[i], ages[i]))

print(zipped)

[('Andy', 29), ('Lisa', 24), ('Dayton', 50)]


Imagine having to write such a snippet of code every time we need to do something with elements from two lists! As you can imagine, it can cause code to be bloated and distracts from the main logic of the program. 

Here's how `zip` is typically used in a program.  

In [7]:
for tup in zip(my_colleagues, ages):
    name = tup[0]
    age = tup[1]
    print("%s's age is %d" % (name, age))

Andy's age is 29
Lisa's age is 24
Dayton's age is 50


In fact, we can do even better in terms of readibility. We can utilize what is known as **list unpacking** to rewrite this for loop. 

In [8]:
for name, age in zip(my_colleagues, ages): # The syntax name, age is what is known as list unpacking
    print("%s's age is %d" % (name, age))

Andy's age is 29
Lisa's age is 24
Dayton's age is 50


Of course, `zip` is used in many other context other than to simplify `for` loops. Can you think of any other situations where you might need to use `zip`? 

## 11.2 `enumerate`

As the name of this function suggests, `enumerate` is useful when we wish to produce a count of items in the list. This is one of the most useful functions you will ever use in Python. It's utility can not be understated. 

`enumerate` works by producing from a list $$ x_0, x_1, \ldots, x_n$$ the following list of tuples (note the 0 indexing) $$(0, x_0), (1, x_1), \ldots, (n, x_n) $$.  

We can use list unpacking to capture both the enumerated index and the object itself. You will use `enumerate` most often in `for` loops. Below is an example where we wish to assign staff names to staff id numbers based on a running serial number. 

In [9]:
staff_id = dict()

for i, name in enumerate(my_colleagues):
    id_no = 's2017-'+str(i) # The str function coerces and integer i into 'i'
    staff_id[id_no] = name

print("A list of staff id numbers")    
print(staff_id.keys())
print("and the respective staff names")
print(staff_id.values())

A list of staff id numbers
['s2017-1', 's2017-0', 's2017-2']
and the respective staff names
['Lisa', 'Andy', 'Dayton']


## 11.3 List comprehension

One one the great things about `for` loops in Python is there are easy to write and understand. However, this comes at a cost: time. `for` loops in Python are slow. Thus, if the code is iterated over a large number of loops, it *will* take time. 

Let's see this. The scenario here is that we want to assign a running serial number to staff id. This will involve coercing `int` into `str` type. The thing is, we have `5000` staff. So let's see how much computer time it takes by using `%timeit`. 

In [1]:
%%timeit

serial_numbers = list()
for i in range(0,5000): # 5000 staff, so we need 5000 int's
    serial_numbers.append('s'+str(i)) # our serial numbers are prefixed with 's'

1.87 ms ± 10.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Notice that the entire script needed about 1.8 ms to execute. This isn't exactly a short amount of time as far as computers go. Just imagine that we have to do this for 10 times in a row! 

Fortunately Python implements what is known as **list comprehension** which is a way of writing `for` loops in a more compact and abstracted way. If we think about a `for` loop to create a list element by element, the code will look something like this: 

    for i in iterable:
        
        do code and return result_i
        append(result_i) to list[i]

Python **list comprehension** provides an alternative syntax which does the exact same thing but faster. The syntax is:

    [do code for i in iterable]

The code is enclosed in `[ ]` because we are using **list comprehension** to create a `list` using a set instruction for each item in the iterable (think of iterables as a `list`). If one is familiar with set notation from calculus courses, list comprehension is syntatically similiar to the following $$\{\, f(x_n) \mid x_n \in A\}$$ where $f$ is some function meant to be evaluated element-wise on each element in a set $A$.  

Now let's refactor the for loop above using list comprehension and time the script. 

In [4]:
%%timeit

serial_numbers = ['s'+str(i) for i in range(5000)]

1.68 ms ± 11 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


That's an improvement of about 10.2 %! 

Let's see another example to really familiarize ourselves with list comprehensions. In the following example, we wish to to extract the first three letters of the months in a year and capitalize them 

In [8]:
months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", 
          "November", "December"]

# multiline statements are allowed in Python as long as they are enclosed in some sort of braces. 

short_name = []
mk_list = short_name.append # Here's a neat trick, assign the append method to a variable mk_list.
                            # mk_list is now a function

for month in months:
    mk_list(month[0:3].upper()) # .upper() is a string method that simply capitalizes all letters in a string. 
    
print(short_name)


['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC']


To refactor this into a list comprehension statement, we first identify the code that is being looped over. That is 

    mk_list(month[0:3].upper())

However, this composite statement can be broken down into steps: 

1. `month[0:3]` just extracts the first 3 letters from `month`

1. Calling the `.upper()` method on the string of 3 letters captilizes them. 

1. Calling `mk_list` is essentially the task of appending the result of the previous two steps to the list `short_name`. 

In list comprehension, the last step is taken care of. Thus, the essential part of the code is `month[0:3].upper()`. Now we identify the iterable: This is simply the list `months` (note plural). 

What is the variable to indicate the particular month as we iterate over the lists of months? This is simply denoted by the name `month` (note singular). There is nothing particularly special about using `month` to name each object in the list `months`. I could have easily used `mon` as well. In that case, the essential part of the code which is being looped should be written `mon[0:3].upper()`. With that clarified, the list comprehension statement is

In [9]:
short_name = [month[0:3].upper() for month in months]
print(short_name)

['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC']


# 12. A concluding demonstration
_____
In this section I want to pull together all that we have learnt in this unit to write a script that illustrates and implements the [RSA](https://en.wikipedia.org/wiki/RSA_(cryptosystem) cryptosystem method. While I will not go through the mathematical details of RSA, I will describe an outline of the program. The script is an attempt to simulate how  Bob can pass sensitive information (like a password) securely to Alice over and open channel using public key cryptography. 

A brief sypnosis on what you will see. 

1. Although we did not cover this in the course, I have created a class called `RSA` which will be used to create an object called `rsa`. Alice will use the *instance method* `rsa.keygen()` which is a `function` to create a pair of tuples: a public and private key. 

1. Bob will use Alice's public key, a pair numbers $(e, n)$ to encrypt his message by the following way: If $M$ is his message coded as a number ($ < n$), the ciphertext $C(M)$ is defined to be $$ C(M) \equiv M^e\mod n $$ This is implemented in Python as `pow(M, e, n)` where `pow` is another of Python's built in function to implement the following code `(M**e)%n` without running into overflow problems.   

1. Bob sends the ciphertext $C(M)$ to Alice to be decrypted. $M$ can be recovered from $C(M)$ by the following $$M \equiv C(M)^d \mod n $$ where $d$ is Alice's private key exponent which *must not* be revealed to anybody else. 

1. The public key exponent is usually a large (but not to big) prime number which does not divide $n$. However, $n$ itself is very huge (almost 200 digits long). 

Note that this is merely a simulated exchange and nothing here is meant to be a reflection of how cryptosystems are actually implemented in real life. But the essential ideas of the algorithm are real and I want to use this to illustrate program flow and how Python scripts will generally look like. 

Note too that the code for the `RSA` class is stored on a  `RSA.py` file. So you will get to see how to import user written code into Jupyter Notebook. 

In [4]:
"""
Before any public key encryption can be carried out. Alice must generate a pair of numbers known as the public
and private exponents. She can give the public key to Bob for encryption but must keep the private key secret. 
"""

from RSA import RSA
import sys

rsa = RSA()

rsa.keygen(100) 
e, n = rsa.get_public_key()

print("Alice's public key exponent and modulus which will be passed to Bob")
print("(e, n) = ({}, {})".format(e, n))

#print("WARNING! Remember this number. Variable is set to be deleted after display!")
#print("{}".format(rsa.d))
d = rsa.d

#print(n > sys.maxsize)

(2504750317581791448985588752089058367438510412639011327498285803746935669168501483003807414025965308, 0, 1)
(1922518107136737381672168164299776692285333654030003069900854772943687487541073642122567353124593338, 1, -1)
(582232210445054067313420587789281675153176758609008257597431030803248181627427840881240060901371970, -1, 2)
(175821475801575179731906400931931666825803378202978297108561680533942942658790119478847170420477428, 4, -7)
(54767783040328528117701384993486674675766624000073366271745989201419353651057482444698549639939686, -13, 23)
(11518126680589595378802245951471642798503506202758198293323712929684881705617672144751521500658370, 43, -76)
(8695276317970146602492401187600103481752599189040573098451137482679826828586793865692463637306206, -185, 327)
(2822850362619448776309844763871539316750907013717625194872575447005054877030878279059057863352164, 228, -403)
(226725230111800273562866895985485531499878147887697513833411141664662197494159028515290047249714, -869, 1536)
(102147601

In [5]:
""" 
Here, Bob recieves Alice's public key (e,n) and proceeds to encrypt his message. He must first convert his 
message into a number M < n. Then he calculates C(M) by (M**e)%n. C(M) is the ciphertext send to Alice. 

In the script below, we create M by using a dictionary to translate between text to M. 
"""
DEFAULT_BLOCKSIZE = 128

if sys.version_info > (3,): # Python 2 and Python 3 compatibility
    raw_input = input
    
s = raw_input("Input Bob's message (max 128 characters)\n>> ")

m = bytearray(s, "ascii")
block = 0
for i, unit in enumerate(m):
    block += unit*(DEFAULT_BLOCKSIZE**i)
print("Numerical representation of string by ASCII encoding")
print(block)

assert block < n

C_M = pow(block, e, n)

print("Cipher Text")
print(C_M)
#print(any([num > sys.maxsize for num in C_M]))

Input Bob's message (max 128 characters)
>> howdy mate
Numerical representation of string by ASCII encoding
939974343295487571944
Cipher Text
11045298581390720201751284965884390719413469440397662764051563719588160958339009784612175984534153336220588361292614908317863428389865699819001108132200756811976876929250798831590585119213065798798874


In [6]:
""" 
Once Alice recieves the message she can decipher it using the formula (C(M)**d)%n
"""
M = pow(C_M, d, n)

assert M == block

text = list()
i = 0

while M//128**i > 0:
    i += 1

for power in range(i)[::-1]: 
    q, M = divmod(M, 128**power)
    text.insert(0, q)

print(text)

assert all([num < 128 for num in text])

text = bytes(text)
text = text.decode("ascii")

print("Decoded message")
print(text)

[104, 111, 119, 100, 121, 32, 109, 97, 116, 101]
Decoded message
howdy mate


## Looking forward 

The horizon of Python programming stretches ever onward! This course represents just  a glimpse of what you can do with Python. In fact, in the previous demo, I wrote the encryption script using the `class` keyword which is the way Python creates objects. When you start using `pandas` and `matplotlib`, you will be interacting with objects the same way I showed you how in the script above. In `pandas`, the fundamental object you will be using is the `DataFrame` while in `matplotlib`, it is `Figure` and `Axes`. 

With what you have learnt today, you will be able to tackle the challenges of learning how to deal with data using `pandas` and visualize it using `matplotlib` or `seaborn`. Things that we did not cover in this course are: 

   * Object Oriented Programming. The `class` keyword, object inheritance and methods. 
   * Iterators and generators. The `yield` keyword. 
   * Decorators and decorator statements. Decorators are used in IPython interactive widgets and web development frameworks. 
   * Module and packages.
   * Unit testing. The `unittest` module. 
   * Debugging and code profiling. 

The knowledge you have learnt here is enough to enable you to explore these topics even further. All the best!